src/app/clusters/occupancy-sensor-server/OccupancySensingCluster.cpp

```diff
@@ -0,0 +1,333 @@
+        { mHasHoldTime && mShowDeprecatedAttributes && mFeatureMap.Has(Feature::kUltrasonic),
+          Attributes::UltrasonicOccupiedToUnoccupiedDelay::kMetadataEntry },
+        { mHasHoldTime && mShowDeprecatedAttributes && mFeatureMap.Has(Feature::kPhysicalContact),
+          Attributes::PhysicalContactOccupiedToUnoccupiedDelay::kMetadataEntry },
+    };
+
+    return listBuilder.Append(Span(OccupancySensing::Attributes::kMandatoryMetadata), Span(optionalAttributes));
+}
+
+DataModel::ActionReturnStatus OccupancySensingCluster::SetHoldTime(uint16_t holdTime)
+{
+    VerifyOrReturnError(mHasHoldTime, Protocols::InteractionModel::Status::UnsupportedAttribute);
+    VerifyOrReturnError(holdTime >= mHoldTimeLimits.holdTimeMin, Protocols::InteractionModel::Status::ConstraintError);
+    VerifyOrReturnError(holdTime <= mHoldTimeLimits.holdTimeMax, Protocols::InteractionModel::Status::ConstraintError);
+
+    if (mHoldTime == holdTime)
+    {
+        return DataModel::ActionReturnStatus::FixedStatus::kWriteSuccessNoOp;
+    }
+
+    mHoldTime = holdTime;
+
+    // If a timer is currently active, we need to adjust its duration to reflect the new hold time.
+    if (mTimerDelegate && mTimerDelegate->IsTimerActive(this))
+    {
+        // Calculate the time that has already elapsed since the timer was started.
+        System::Clock::Timestamp now                = mTimerDelegate->GetCurrentMonotonicTimestamp();
+        System::Clock::Timestamp elapsedTime        = now - mTimerStartedTimestamp;
+        System::Clock::Seconds16 newHoldTimeSeconds = System::Clock::Seconds16(mHoldTime);
+
+        // If the elapsed time is already greater than or equal to the new hold time,
+        // the timer should have already fired. Cancel the current timer and immediately
+        // set the state to unoccupied.
+        if (elapsedTime >= newHoldTimeSeconds)
+        {
+            mTimerDelegate->CancelTimer(this);
+            DoSetOccupancy(false);
+        }
+        else
+        {
+            // Otherwise, restart the timer for the remaining duration.
+            System::Clock::Timeout remainingTime = newHoldTimeSeconds - elapsedTime;
+            mTimerDelegate->StartTimer(this, remainingTime);
+        }
+    }
+
+    if (mDelegate)
+    {
+        mDelegate->OnHoldTimeChanged(holdTime);
+    }
+
+    if (mContext != nullptr)
+    {
+        // There's not much we can do if persistence fails here, so we ignore the return value.
+        // The application must ensure that persistence is working correctly.
+        RETURN_SAFELY_IGNORED mContext->attributeStorage.WriteValue(
+            { mPath.mEndpointId, OccupancySensing::Id, Attributes::HoldTime::Id },
+            { reinterpret_cast<const uint8_t *>(&mHoldTime), sizeof(mHoldTime) });
+    }
+
+    return Protocols::InteractionModel::Status::Success;
+}
+
+void OccupancySensingCluster::SetOccupancy(bool occupied)
+{
+    if (mHasHoldTime && mTimerDelegate)
```

Comment:
It seems that mHasHoldTime ==true is implicit of mTimerDelegate, shouldn't we have a VerifyOrDie to prevent having a cluster that pretends to support HoldTime but doesn't have a TimerDelegate?

Or should this be part of an Init() method that would fail to indicate invalid configuration?

We probably don't want a device on a network that pretends to support HoldTime without a timer.